PT-FIXES: DOS-1290 DateParser in java

abdelaziz-mahdy opened the pull request.

abdelaziz-mahdy committed:
Refactor DateUtil to use grammar-based DateParser for date and datetime parsing

- Updated DateUtil.js to replace legacy date parsing logic with DateParser for improved accuracy and maintainability.
- Modified tests in DateUtilJSTest.js and DateUtilTest.js to reflect changes in date parsing behavior, ensuring exceptions are thrown for unsupported formats.
- Adjusted test cases to validate normalization of invalid dates instead of throwing exceptions.
- Added DateParserJavaTest to the test suite for comprehensive coverage of date parsing functionality.
- Updated pom.js to include DateParser and DateParserJavaTest in the build configuration.

abdelaziz-mahdy committed:
feat(DateParser): enhance date parsing to allow 00-99 for minutes/seconds and improve error handling for invalid formats

abdelaziz-mahdy committed:
Merge with dev

abdelaziz-mahdy committed:
refactor(pom): remove unused SearchView parser from POM configuration

jlhughes commented on src/foam/parse/DateParser.js:
```diff
@@ -829,72 +802,71 @@ foam.CLASS({
-        }
-        if ( result.second !== undefined && (result.second < 0 || result.second > 59) ) {
-          return this.validateDate(this.INVALID_DATE, str);
+        if ( debugLog ) {
```
Will this remain in the final commit?

abdelaziz-mahdy replied:
not, that was missed from myside, removed, and thank you for noticing it

abdelaziz-mahdy committed:
refactor(DateParser): remove debug logging from parseDateTime method

abdelaziz-mahdy committed:
merge with dev

jlhughes commented on src/foam/parse/DateParser.java:
```diff
@@ -0,0 +1,1079 @@
+
+  /**
+   * Parse a datetime string using UTC time.
+   * Uses time if present, otherwise sets to midnight.
```
Reason for midnight over noon? We default Date (no time) to noon elsewhere in foam.

abdelaziz-mahdy replied:
only for the case of DateTimeUTC, since its not affected by timezone, so need to set it for noon

jlhughes commented on src/foam/parse/DateParser.java:
```diff
@@ -0,0 +1,1079 @@
+  private int parseMonthName(String monthName) {
+    String month = monthName.toUpperCase();
+    switch (month) {
+      case "JAN": return 0;
```
If concerned about locale, this could be
return java.time.format.DateTimeFormatter.ofPattern("MMM")
      .withLocale(Locale.ENGLISH)
      .parse(monthName)
      .get(ChronoField.MONTH_OF_YEAR);

abdelaziz-mahdy replied:
currently front and backend, dont check the locale, I believe it should be future improvements since I should make both in sync

kgrgreer commented on src/foam/parse/DateParser.java:
```diff
@@ -0,0 +1,1079 @@
+      Map<String, Object> resultMap = (Map<String, Object>) parseResult.value();
+
+      // Determine if this is a datetime or date-only result
+      if ( resultMap.containsKey("hour") || resultMap.containsKey("minute") || resultMap.containsKey("second") ) {
```
Storing in a Map is going to be expensive because it needs to be allocated and then garbage collected.

abdelaziz-mahdy replied:
done, migrated both client and server codes

kgrgreer commented on src/foam/parse/DateParser.java:
```diff
@@ -0,0 +1,1079 @@
+    }
+
+    try {
+      StringPStream sps = new StringPStream();
```
Why not just new StringPStream(str)?

abdelaziz-mahdy replied:
done

abdelaziz-mahdy committed:
merge with dev

abdelaziz-mahdy committed:
chore: update code structure for improved readability and maintainability

abdelaziz-mahdy committed:
feat: enhance DateParser to support STRING mode for date parsing

abdelaziz-mahdy committed:
fix: update date parsing logic to allow single-digit values for day, hour, minute, and second

abdelaziz-mahdy committed:
Merge With Dev

abdelaziz-mahdy committed:
feat: add strict validation mode for date parsing with error handling

abdelaziz-mahdy committed:
refactor: streamline date parser tests and enable strict validation mode

abdelaziz-mahdy committed:
feat: implement strict and non-strict validation modes for date parsing

abdelaziz-mahdy committed:
fix: update error message for unsupported date format in strict mode

abdelaziz-mahdy committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into date-util-date-parser-in-java

abdelaziz-mahdy committed:
feat(date-util): implement lenient and strict validation modes for date parsing

abdelaziz-mahdy committed:
feat(parse): add new date format tests for Java compatibility

- Implemented tests for various date formats to ensure parity with JavaScript date parsing.
- Added tests for timestamps (both 10-digit and 13-digit), single-digit month/day formats, space-separated month names, MM/DD/YY formats, and fractional seconds.

abdelaziz-mahdy committed:
feat(parse): add Unix Date.toString() format and timestamp support to DateParser

  Add support for parsing Unix/Java Date.toString() format (e.g., "Tue Apr 01 05:17:59 GMT 2025")
  with timezone offset handling. Add Unix timestamps (10-digit seconds, 13-digit milliseconds).
  Make strictValidation static to match JavaScript singleton behavior.

  Remove YYMMDD from main entry point - ambiguous with MMDDYY, use opt_name='yymmdd' explicitly.
